File: kalitech-engine/src/main/resources/kalitech/engine/modules/Mesh/helpers/MeshIds.js
"use strict";

// ------------------- strict assumptions -------------------
// 1) SurfaceHandle provides id():int
// 2) ENGINE.surface().attachedBody(surfaceId)->int
// 3) ENGINE.physics() operates only on bodyId

function surfaceId(handle) {
    if (!handle || typeof handle.id !== "function") throw new Error("[MSH] SurfaceHandle must provide id()");
    const sid = handle.id() | 0;
    if (sid <= 0) throw new Error("[MSH] invalid surfaceId=" + sid);
    return sid;
}

function requireSurface(engine) {
    const s = engine.surface();
    if (!s || typeof s.attachedBody !== "function") {
        throw new Error("[MSH] ENGINE.surface().attachedBody(surfaceId) is required");
    }
    return s;
}

function requirePhysics(engine) {
    const p = engine.physics();
    if (!p) throw new Error("[MSH] ENGINE.physics() is required");
    return p;
}

function resolveBodyId(engine, sid) {
    const s = requireSurface(engine);
    const bid = (s.attachedBody(sid) | 0);
    if (bid <= 0) throw new Error("[MSH] surface has no physics body (bodyId=0)");
    return bid;
}

module.exports = Object.freeze({
    surfaceId,
    requireSurface,
    requirePhysics,
    resolveBodyId,
});

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Mesh/helpers/MeshIds.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
function surfaceId(self, handle)
    if not handle or KTypeOf(handle.id) ~= "function" then
        error(
            LuaConstruct(Error, "[MSH] SurfaceHandle must provide id()"),
            0
        )
    end
    local sid = bit32.bor(
        handle:id(),
        0
    )
    if sid <= 0 then
        error(
            LuaConstruct(
                Error,
                "[MSH] invalid surfaceId=" .. tostring(sid)
            ),
            0
        )
    end
    return sid
end
function requireSurface(self, engine)
    local s = engine:surface()
    if not s or KTypeOf(s.attachedBody) ~= "function" then
        error(
            LuaConstruct(Error, "[MSH] ENGINE.surface().attachedBody(surfaceId) is required"),
            0
        )
    end
    return s
end
function requirePhysics(self, engine)
    local p = engine:physics()
    if not p then
        error(
            LuaConstruct(Error, "[MSH] ENGINE.physics() is required"),
            0
        )
    end
    return p
end
function resolveBodyId(self, engine, sid)
    local s = requireSurface(_G, engine)
    local bid = bit32.bor(
        s:attachedBody(sid),
        0
    )
    if bid <= 0 then
        error(
            LuaConstruct(Error, "[MSH] surface has no physics body (bodyId=0)"),
            0
        )
    end
    return bid
end
M = KObject:freeze({surfaceId = surfaceId, requireSurface = requireSurface, requirePhysics = requirePhysics, resolveBodyId = resolveBodyId})

return M
